### ISSUE TITLE
[BUG] Large performance regression for cuDF / PyArrow conversions (`to_arrow` in particular)

### ISSUE THREAD

**Issue by beckernick** [rapidsai/cudf#17075]
It looks like we had a substantial performance regression in CPU/GPU transfer time in 24.10 (compared to stable v24.08 release) specifically for the PyArrow interfaces (which is separate from `to_pandas`/`from_pandas`). I'm not sure if this limited to the Python Arrow interfaces vs. C++.

I believe we use this sparingly in cuDF pandas, but it's the interface we use for Polars. For queries materializing large tables (e.g., joins), this may be a non-trivial performance impact.

The larger regression is in the GPU -> CPU direction (`to_arrow`), which in the below example is about 80% slower than before when transferring a large dataframe of 64 bit columns (16GB, in this case). 

```
(rapids-24.08) $ ipython interop-benchmark.ipy
to_arrow() time:
4.32 s ± 1.12 ms per loop (mean ± std. dev. of 2 runs, 2 loops each)
from_arrow() time:
1.74 s ± 36 ms per loop (mean ± std. dev. of 2 runs, 2 loops each)
GPU -> CPU -> GPU Arrow roundtrip time:
6.11 s ± 27 ms per loop (mean ± std. dev. of 2 runs, 2 loops each)
to_pandas() time:
13.3 s ± 20.7 ms per loop (mean ± std. dev. of 2 runs, 2 loops each)
GPU -> CPU -> GPU Pandas roundtrip time:
19.3 s ± 204 ms per loop (mean ± std. dev. of 2 runs, 2 loops each)
```

```
(rapids-24.10) $ ipython interop-benchmark.ipy
to_arrow() time:
7.77 s ± 3.46 ms per loop (mean ± std. dev. of 2 runs, 2 loops each)
from_arrow() time:
2.3 s ± 435 ms per loop (mean ± std. dev. of 2 runs, 2 loops each)
GPU -> CPU -> GPU Arrow roundtrip time:
9.54 s ± 119 ms per loop (mean ± std. dev. of 2 runs, 2 loops each)
to_pandas() time:
12.8 s ± 33.8 ms per loop (mean ± std. dev. of 2 runs, 2 loops each)
GPU -> CPU -> GPU Pandas roundtrip time:
19 s ± 175 ms per loop (mean ± std. dev. of 2 runs, 2 loops each)
```

These were run consecutively on the same machine on the same file.

interop-benchmark.ipy:
```python
import cudf
import time

path = "testfile.parquet"

df = cudf.read_parquet(path)
cpu_table = df.to_arrow()

print("to_arrow() time:")
%timeit -n2 -r2 df.to_arrow()

print("from_arrow() time:")
%timeit -n2 -r2 df = cudf.DataFrame.from_arrow(cpu_table)

print("GPU -> CPU -> GPU Arrow roundtrip time:")
%timeit -n2 -r2 cudf.DataFrame.from_arrow(df.to_arrow())

print("to_pandas() time:")
%timeit -n2 -r2 df.to_pandas()

print("GPU -> CPU -> GPU Pandas roundtrip time:")
%timeit -n2 -r2 cudf.DataFrame.from_pandas(df.to_pandas())
```

Data generation script:
```python
import cudf

NROWS = 200000000
N_INT = 5
N_FLOAT = 5

DTYPES = {f"i{x}": int for x in range(N_INT)}
DTYPES.update({f"f{x}": float for x in range(N_FLOAT)})

df = cudf.datasets.randomdata(nrows=NROWS, dtypes=DTYPES, seed=12)
size_mb = df.memory_usage(deep=True).sum()/1e6

print(f"{size_mb} MB dataframe")

path = "testfile.parquet"
df.to_parquet(path)
```

Environments:
- Stable RAPIDS v24.08 conda environment
- Nightly RAPIDS v24.10 with cuDF on build a399 at commit 742eaadb92b0c5159d92be49e647a17e8c1d0b9b

**Comment by vyasr** [comment]
The `from_arrow` regression seems to be just noise. I was not able to reproduce it on reruns. The `to_arrow` regression is definitely real and is being fixed in #17097.

**Comment by vyasr** [comment]
Benchmarks are in https://github.com/rapidsai/cudf/pull/17081#issuecomment-2415332586.
